base/src/main/java/com/mindera/skeletoid/widgets/tooltip/PopupTooltip.kt

```diff
@@ -296,75 +213,101 @@ class PopupTooltip(
+        //Setup the arrow
+        arrow?.let { nonNullArrow ->
+            setupArrow(nonNullArrow, layout, context, screenPos[0])
         }
 
         setupTooltipBehaviour(context, layout, tooltipWidth)
 
-        val x: Float = screenPos[0] - tooltipWidth / 2f - getValueInPixels(horizontalMargin, context).toInt() + anchorView.width / 2f
-        val y: Float = screenPos[1] - estimatedTooltipHeight - getValueInPixels(verticalMargin, context).toInt() - (if (arrow) tooltipArrowHeight else 0)
-
-        //Handle arrow
-        if (arrow) setupArrow(layout, context, tooltipArrowWidth, screenPos[0])
-
-        //Handle timed dismiss
-        if (!indefinite) {
-            handler.postDelayed({ clearTooltip() }, showDuration)
-        }
-
-        showTooltipAtLocation(x, y, displayFrame.top, screenPos[0], tooltipWidth, context)
+        val (initialX, initialY) = positionDelegate.calculateTooltipInitialPosition(
+            screenPos = screenPos,
+            tooltipWidth = tooltipWidth,
+            estimatedTooltipHeight = estimatedTooltipHeight,
+            anchorView = anchorView,
+            arrow = arrow,
+            tooltipGravity = tooltipGravity,
+            horizontalMargin = horizontalMargin,
+            verticalMargin = verticalMargin,
+            context = context
+        )
+
+        return positionDelegate.calculateTooltipFinalPosition(
+            x = initialX,
+            y = initialY,
+            statusBarHeight = displayFrame.top,
+            anchorXScreenPos = screenPos[0],
+            tooltipWidth = tooltipWidth,
+            anchorView = anchorView,
+            tooltipGravity = tooltipGravity,
+            horizontalMargin = horizontalMargin,
+            horizontalMarginClipped = horizontalMarginClipped,
+            verticalMargin = verticalMargin,
+            verticalMarginClipped = verticalMarginClipped,
+            context = context
+        )
     }
 
     /**
      * If the arrow flag is set to true, this method enables the arrow and places it at the
      * required place, at the top center of the anchor view.
      */
-    private fun setupArrow(layout: View, context: Context, tooltipArrowWidth: Float, anchorViewAbsoluteX: Int) {
-        if (arrowBackgroundId != null && arrowStrokeId != null) {
-            val arrow = layout.findViewById<ImageView>(arrowBackgroundId ?: 0)
-            val stroke = layout.findViewById<ImageView>(arrowStrokeId ?: 0)
-            val elevation = getValueInPixels(elevation, context)
-
-            arrowDrawable?.let {
-                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
-                    arrow.setImageDrawable(context.getDrawable(it))
-                    arrow.elevation = elevation
-                    stroke.setImageDrawable(context.getDrawable(it))
-                    stroke.elevation = elevation
-                }
-            }
+    private fun setupArrow(
+        arrow: ArrowData,
+        layout: View,
+        context: Context,
```

Comment:
Do we need to receive context? Can't we get it from the layout view?

Reply:
Yeah, we can 🤦